lib/dynamodb/dynamodb_kv.c

```diff
@@ -65,6 +65,138 @@ dynamodb_kv_put(const char * table, const char * key,
 	return (s);
 }
 
+/**
+ * dynamodb_kv_icas(table, key, buf, len, buf2, len2):
+ * Construct a DynamoDB request body for an ICAS replacing V=${buf} (of length
+ * ${len}) with V=${buf2} (of length ${len2}), associated with K=${key} in
+ * DynamoDB table ${table}.
+ */
+char *
+dynamodb_kv_icas(const char * table, const char * key,
+    const uint8_t * buf, size_t len, const uint8_t * buf2, size_t len2)
+{
+	/**
+	 * An ICAS is a PutItem request with a body (with spaces added) of:
+	 * { "TableName": "TABLE",
+	 *   "Item": {
+	 *     "K": { "S": "KEY" },
+	 *     "V": { "B": "BASE64NEW" }
+	 *   },
+	 *   "ConditionExpression": "(V = :old) OR (V = :new)",
+	 *   "ExpressionAttributeValues": {
+	 *       ":old": { "B": "BASE64OLD" },
+	 *       ":new": { "B": "BASE64NEW" }
+	 *   },
+	 *   "ReturnConsumedCapacity": "TOTAL"
+	 * }
+	 */
+	const char * s1 = "{\"TableName\":\"";
+	const char * s2 = "\",\"Item\":{\"K\":{\"S\":\"";
+	const char * s3 = "\"},\"V\":{\"B\":\"";
+	const char * s4 = "\"}},\"ConditionExpression\":\"(V=:old)OR(V=:new)\",";
+	const char * s5 = "\"ExpressionAttributeValues\":{\":old\":{\"B\":\"";
+	const char * s6 = "\"},\":new\":{\"B\":\"";
+	const char * s7 = "\"}},\"ReturnConsumedCapacity\":\"TOTAL\"}";
+	size_t slen = strlen(s1) + strlen(table) + strlen(s2) + strlen(key) +
+	    strlen(s3) + ((len2 + 2) / 3) * 4 + strlen(s4) + strlen(s5) +
```

Comment:
Umm.  I wonder if it's worth doing something like
```
#define BASE64LEN(len) (((len + 2) / 3 * 4)
```

in `b64encode.h` so that these lines could be:

```srtlen(s3) + BASE64LEN(len2) + strlen(s4) + strlen(s5) +
BASE64LEN(len) + strlen(s6) + BASE64LEN(len2) + strlen(s7)
```

Also, it might be worth adding a comment about "yes, we need to use len2 twice, because BASE64NEW appears twice in the request".

Reply:
after checking `b64encode.h`, the macro name should obviously be `b64len`, not `BASE64LEN`.

Reply:
Sure.  Please open a separate PR in libcperciva for adding the macro to `b64encode.h` and then (after this PR has been merged) a PR here which makes all the necessary changes (not just for the newly added ICAS and CREATE operations).